### ISSUE TITLE
replicator seems lead to high cpu when no any message publish to liftbridge

### ISSUE THREAD

**Issue by hl3w22bupt** [liftbridge-io/liftbridge#313]
I created a stream `testtlblog` with 2 partitions and replica-factor is 2, after running for a while, I stopped the producer, and the liftbridge server and nats server still running. But some times later, I found high cpu occupancy with `top` command
![image](https://user-images.githubusercontent.com/3297112/103861620-ae38c680-50f8-11eb-8c0e-1bf11102c390.png)

And also I found massive nats communication is ongoing between liftbridge servers, **look likes they are doing some job about replica**.  
> [#25689] Received on [testlb.testtlblog.0.replicate]: '�C�
> server-2���'
> [#25690] Received on [_INBOX.FRbHzl6QRyqq2LNMOKbW1m.uHyUJJr7]: '�C!]��'
> [#25691] Received on [testlb.notify.server-2]: '�C�
> 
> testtlblog'
> [#25692] Received on [testlb.testtlblog.0.replicate]: '�C�
> server-2���'
> [#25693] Received on [_INBOX.FRbHzl6QRyqq2LNMOKbW1m.vGjesjOE]: '�C!]��'
> [#25694] Received on [testlb.notify.server-2]: '�C�
> 
> testtlblog'
> [#25695] Received on [testlb.testtlblog.0.replicate]: '�C�
> server-2���'
> [#25696] Received on [_INBOX.FRbHzl6QRyqq2LNMOKbW1m.crJLwdFT]: '�C!]��'
> [#25697] Received on [testlb.notify.server-2]: '�C�

I got the above infomartion by `nats-sub` tool. `testlb` is the clustering namespace, and `testtlblog` is the stream name

These log messages have already been cleaned when retention max age reached, how it happened? I need your help and will appreciate very much

> root /opt/liftbridge# ls -l data/server-2/streams/testtlblog/0/
> total 8
> -rw-r--r-- 1 root root 10485760 Jan  6 18:36 00000000000559806923.index
> -rw-r--r-- 1 root root        0 Jan  6 18:36 00000000000559806923.log
> -rw------- 1 root root       17 Jan  6 18:44 leader-epoch-checkpoint
> -rw------- 1 root root        9 Jan  7 14:43 replication-offset-checkpoint

By the way, my liftbridge server is v1.5.0, golang version is 1.15.5

**Comment by hl3w22bupt** [comment]
FYI, When I resumed to publishing message right now, it is surprisingly the symptom disappear, the cpu load decrease, everything becomes normal. **Again I stopped the publishing, it never reoccurs so far,** cpu load is still low, I am not sure whether it will reoccur again, I will keep the servers running and keep following up

**Comment by ekbfh** [comment]
Have you tried disabling compression?

**Comment by tylertreat** [comment]
That is interesting. Liftbridge is designed to specifically idle when there is no new data published to a partition and replicas are caught up. What's strange is `testlb.notify.server-2` is the NATS subject used to signal to the replica that new data is available on the partition and that the replica should wake up to replicate it. I wouldn't expect to see messages on this subject unless there are still new messages being published to that partition.

**Comment by hl3w22bupt** [comment]
> Have you tried disabling compression?

Do you mean the configuration item `compact`? 

Here is my configure file for server-1.

> ---
> listen: localhost:9292
> host: localhost
> port: 9292
> data.dir: data/server-1
> 
> # Configure logging.
> logging:
>   level: debug
>   recovery: false
>   raft: false
> 
> # Define NATS cluster to connect to.
> nats.servers:
>   - nats://localhost:4422
>   - nats://localhost:4433
> 
> # Specify message stream settings.
> streams:
>   retention.max:
>     age: 24h
>     messages: 10000
>   compact.enabled: false
>   cleaner.interval: 1m
> 
> # Specify cluster settings.
> clustering:
>   server.id: server-1
>   namespace: testlb
>   raft.bootstrap.seed: true
>   replica.max.lag.time: 20s
>   #min.insync.replicas=1
> 
> cursors.stream.partitions: 1

**Comment by hl3w22bupt** [comment]
> That is interesting. Liftbridge is specifically designed to idle when there is no new data published to a partition and replicas are caught up. What's strange is `testlb.notify.server-2` is the NATS subject used to signal to the replica that new data is available on the partition and that the replica should wake up to replicate it. I wouldn't expect to see messages on this subject unless there are still new messages being published to that stream. Is one of the partitions still being published to when this occurs?

Yes, It confused me so much, and I still could not figure out although tried to read the code. 

I double checked, it seems that no any message published through nats and liftbridge.

By the way, I always publish message through nats in this case

**Comment by tylertreat** [comment]
@hl3w22bupt You are certain no other messages are being published to the NATS subjects that the stream partitions are subscribed to? My best guess is that there are still messages being received on the stream, which causes the `notify` messages.

Are you able to reproduce this behavior? That would help me debug the issue.

**Comment by hl3w22bupt** [comment]
> @hl3w22bupt You are certain no other messages are being published to the NATS subjects that the stream partitions are subscribed to? My best guess is that there are still messages being received on the stream, which causes the `notify` messages.
> 
> Are you able to reproduce this behavior? That would help me debug the issue.

Can not reproduce now, I will collect more information if it happens again. Thanks so much @tylertreat

**Comment by aversant** [comment]
Reproducing need more time (about 8 hours):
![image](https://user-images.githubusercontent.com/18608678/106361982-1b2f2d00-6342-11eb-8fb1-c96f6e781f11.png)

1. Cleanup `data` folder
2. Create 2 stream with 2 partitions
3. Subscribe on each 4 partition
4. Produce message to one of stream (I send 1 message per second)
5. Wait about 2-8 hours

![image](https://user-images.githubusercontent.com/18608678/106362121-fd15fc80-6342-11eb-8126-c3641d62b932.png)


`liftbridge.yml`

```yml
clustering:
    min.insync.replicas: 2
    raft.bootstrap.peers:
    - mstage-db
    - mstage-wrk
    - mstage-wrk2
    server.id: mstage-db
cursors:
    stream.auto.pause.time: 0
    stream.partitions: 1
data:
    dir: /var/lib/liftbridge
host: 10.10.10.19
logging:
    level: debug
    raft: true
nats.servers:
- nats://10.10.10.19:4222
- nats://10.10.10.20:4222
- nats://10.10.10.21:4222
streams:
    compact.enabled: true
    retention.max:
        age: 1h

```

**Comment by tylertreat** [comment]
Thanks, will investigate.

**Comment by aversant** [comment]
Some fixes on steps:
1. Cleanup data folder
2. Create 2 stream with 2 partitions
3. Subscribe on each 4 partition
4. Produce message to one of stream (I send 1 message per second)
5. **Put one message to another stream**
6. Wait about 2-8 hours

Problem reproduce after `level=debug msg="Appending new log segment for /var/lib/liftbridge/streams/test.stream2/0 with base offset 1"` message. (for my settings after 1 hour)

**Comment by aversant** [comment]
Additional investigation. Add some debug messages to `Liftbridge` code [lb_git_diff1.txt](https://github.com/liftbridge-io/liftbridge/files/6004804/lb_git_diff1.txt). On [leader_epoch_cache.ClearEarliest](https://github.com/liftbridge-io/liftbridge/blob/76757fe795e8a562f523cced37434d4feef660f5/server/commitlog/leader_epoch_cache.go#L129).

(Stream create with `req.segmentMaxAge.value = 3600000 (1h)`
```
# Brokers (3)
ID                   | HOST:PORT
mstage-db            | 10.*.*.*:9292
mstage-wrk2          | 10.*.*.*:9292
mstage-wrk           | 10.*.*.*:9292
# Streams
  ## Name: test.stream2 Subject: test.stream2
    ### Partition: 0
    Leader        : mstage-wrk
    Replicas      : mstage-db, mstage-wrk
    ISR           : mstage-db, mstage-wrk
    HighWatermark : 0
    NewestOffset  : 0
    ### Partition: 1
    Leader        : mstage-wrk
    Replicas      : mstage-db, mstage-wrk
    ISR           : mstage-db, mstage-wrk
    HighWatermark : -1
    NewestOffset  : -1
```

Diff between leader (mstage-wrk) and replica (mstage-db) - call param on ClearEarliest method:

**Leader**: `"ClearEarliest Start%!(EXTRA int64=0, int64=0)"`
**Replicat**: `"ClearEarliest Start%!(EXTRA int64=-1, int64=0)"`

Logs:

```
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg=checkAndPerformSplit
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="Appending new log segment for /var/lib/liftbridge/streams/test.stream2/0 with base offset 1"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="Cleaning log /var/lib/liftbridge/streams/test.stream2/1 based on retention policy {Bytes:0 Messages:0 Age:24h0m0s}"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="Finished cleaning log /var/lib/liftbridge/streams/test.stream2/1"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="Update the leader epoch offset cache to account for deleted segments"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="leaderEpochCache.ClearEarliest%!(EXTRA int64=0)"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="ClearEarliest Start%!(EXTRA int64=0, int64=0)"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="End Clean%!(EXTRA <nil>)"
Feb 18 20:32:15 mstage-wrk liftbridge[20594]: time="2021-02-18 20:32:15" level=debug msg="api: FetchMetadata []"


Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="api: FetchMetadata []"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg=checkAndPerformSplit
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="Appending new log segment for /var/lib/liftbridge/streams/test.stream2/0 with base offset 1"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="Cleaning log /var/lib/liftbridge/streams/test.stream2/1 based on retention policy {Bytes:0 Messages:0 Age:24h0m0s}"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="Finished cleaning log /var/lib/liftbridge/streams/test.stream2/1"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="Update the leader epoch offset cache to account for deleted segments"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="leaderEpochCache.ClearEarliest%!(EXTRA int64=0)"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="ClearEarliest Start%!(EXTRA int64=-1, int64=0)"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="ClearEarliest Loop%!(EXTRA []*commitlog.epochOffset=[])"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="LEN Earlest%!(EXTRA int=0)"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="End Clean%!(EXTRA <nil>)"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="api: FetchMetadata []"
Feb 18 20:32:15 mstage-db liftbridge[18737]: time="2021-02-18 20:32:15" level=debug msg="api: FetchMetadata []"

```

**Comment by tylertreat** [comment]
@aversant Just to clarify, are you seeing the issue after some time _while publishing 1 msg/sec_ or after stopping the publisher and waiting for a while?

**Comment by aversant** [comment]
> @aversant Just to clarify, are you seeing the issue after some time _while publishing 1 msg/sec_ or after stopping the publisher and waiting for a while, i.e. no publishes happening?

I see a problem after an hour (segmentMaxAge interval) if there is one message in the partition. At least in the test it was reproduced under such conditions.  The presence of the partition *while publishing 1 msg/sec* is not required.

**Comment by tylertreat** [comment]
Thanks, so if I understand correctly to reproduce I should:

1. Set segment.max.age to 1hr
2. Create a stream partition with replicationFactor = 2
3. Publish 1 message to partition
4. Wait >1hr

**Comment by aversant** [comment]
> Thanks, so if I understand correctly to reproduce I should:
> 
> 1. Set segment.max.age to 1hr
> 2. Create a stream partition with replicationFactor = 2
> 3. Publish 1 message to partition
> 4. Wait >1hr

Yes, That correct.

**Comment by tylertreat** [comment]
So far I've been unable to reproduce unfortunately. Will continue trying...

**Comment by aversant** [comment]
> So far I've been unable to reproduce unfortunately. Will continue trying...

I got the following:

High CPU usage happens when LB spam message to NATS:
![image](https://user-images.githubusercontent.com/18608678/109039468-a657d480-76ee-11eb-82d9-da798398583f.png)

```
Code when generate message: [Replicator caughtUp](https://github.com/liftbridge-io/liftbridge/blob/76757fe795e8a562f523cced37434d4feef660f5/server/replicator.go#L259). When create segment we have:
`Feb 24 20:05:47 mstage-wrk liftbridge[19222]: time="2021-02-24 20:05:47" level=debug msg="Check if we're caught up%!(EXTRA int64=-1, int64=-1)"`
```

After create stream and put 1 message (`Position: 144, LastOffset: 0`):

```
Feb 24 20:05:49 mstage-wrk liftbridge[19222]: time="2021-02-24 20:05:49" level=debug msg="Check if we're caught up%!(EXTRA int64=0, int64=0)"
Feb 24 20:05:49 mstage-wrk liftbridge[19222]: time="2021-02-24 20:05:49" level=debug msg="[NotifyLEO] param leo: 0"
Feb 24 20:05:49 mstage-wrk liftbridge[19222]: time="2021-02-24 20:05:49" level=debug msg="[NotifyLEO] Segment params. Position: 144, LastOffset: 0, MaxBytes: 268435456"
```

After the passage of time `segment.MaxAge` `Liftbridge` create New segment with `LastOffset: -1`:

```
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg=checkAndPerformSplit
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="Appending new log segment for /var/lib/liftbridge/streams/test.stream2/0 with base offset 1"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="Seal Active Segment"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="[commitlog.Clean] Update the leader epoch offset cache to account for deleted segments"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="[commitlog.Clean] leaderEpochCache.ClearEarliest with segment BaseOffset: %!(EXTRA int64=0)"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="ClearEarliest Start%!(EXTRA int64=0, int64=0)"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="End Clean%!(EXTRA <nil>)"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="Check if we're caught up%!(EXTRA int64=0, int64=0)"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="[NotifyLEO] param leo: 0"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="[NotifyLEO] Segment params. Position: 0, LastOffset: -1, MaxBytes: 268435456"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="Replicator Start GoRoutine%!(EXTRA <-chan struct {}=0xc0005c2fc0)"
Feb 24 20:11:38 mstage-wrk liftbridge[19222]: time="2021-02-24 20:11:38" level=debug msg="Replicator End GoRoutine%!(EXTRA <-chan struct {}=0xc0005c2fc0)"

```

And call method  [r.partition.log.NotifyLEO(r, leo)](https://github.com/liftbridge-io/liftbridge/blob/76757fe795e8a562f523cced37434d4feef660f5/server/replicator.go#L266) -> [NotifyLEO](https://github.com/liftbridge-io/liftbridge/blob/76757fe795e8a562f523cced37434d4feef660f5/server/commitlog/commitlog.go#L625) with params: `[NotifyLEO] param leo: 0`, but new segment created with params: `Position: 0, LastOffset: -1`. With this call [WaitForLEO](https://github.com/liftbridge-io/liftbridge/blob/76757fe795e8a562f523cced37434d4feef660f5/server/commitlog/segment.go#L269) and return on block:

```go
	if s.lastOffset != leo {
		ch := make(chan struct{})
		close(ch)
		return ch
	}
```
This will be repeated until a new message arrives. Then segment `LastOffset` set to 0 and the condition will stop being met and call `return s.waitForData(waiter, s.position)`.

The situation occurs when retention used only time (on bytes the probability is low).

I added some debug to code for logging and set `segment.MaxAge` to `5 m`, for speedup reproduce. Config:

```
streams:
    compact.enabled: false
    retention.max:
        age: 24h
```

Sorry for the  *chaotic description*. I can explain in more detail and attach logs if needed.

**Comment by tylertreat** [comment]
Great detective work. I think the issue is becoming clear. Working on a fix.

**Comment by tylertreat** [comment]
This should be fixed with #329 which is now merged. Please try it and let me know if the issue is resolved. The issue could be quickly reproduced by setting `segment.max.age` to `1m` and `cleaner.interval` to `1m` then publishing 1 message.

**Comment by aversant** [comment]
> This should be fixed with #329 which is now merged. Please try it and let me know if the issue is resolved. The issue could be quickly reproduced by setting `segment.max.age` to `1m` and `cleaner.interval` to `1m` then publishing 1 message.

I installed fix in test environment and Issue is not reproduced. Very thanks.

**Comment by tylertreat** [comment]
Going to create a patch release that will include the fix shortly.
